src/main/groovy/org/ajoberstar/grgit/auth/JschAgentProxySessionFactory.java

```diff
@@ -0,0 +1,186 @@
+    if (con.isPresent()) {
+      IdentityRepository remoteRepo = new RemoteIdentityRepository(con.get());
+      if (remoteRepo.getIdentities().isEmpty()) {
+        logger.info("not using agent proxy: no identities found");
+      } else {
+        logger.info("using agent proxy");
+        jsch.setIdentityRepository(remoteRepo);
+      }
+    } else {
+      logger.info("jsch agent proxy not available");
+    }
+    return jsch;
+  }
+
+  /**
+   * Chooses which agent proxy connector is used.
+   *
+   * @return the connector available at this time
+   */
+  private Optional<Connector> determineConnector() {
+    return Stream.<Supplier<Optional<Connector>>>of(this::sshAgentSelector, this::pageantSelector)
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .findFirst();
+  }
+
+  private Optional<Connector> sshAgentSelector() {
+    try {
+      if (!config.allows(AuthConfig.Option.SSHAGENT)) {
+        logger.info("ssh-agent option disabled");
+        return Optional.empty();
+      } else if (SSHAgentConnector.isConnectorAvailable()) {
+        Optional<USocketFactory> usf = determineUSocketFactory();
+        if (usf.isPresent()) {
+          logger.info("ssh-agent available");
+          return Optional.of(new SSHAgentConnector(usf.get()));
+        } else {
+          logger.info("ssh-agent not available");
+          return Optional.empty();
+        }
+      } else {
+        logger.info("ssh-agent not available");
+        return Optional.empty();
+      }
+    } catch (Throwable e) {
+      logger.info("ssh-agent could not be configured: {}", e.getMessage());
+      logger.debug("ssh-agent failure details.", e);
+      return Optional.empty();
+    }
+  }
+
+  private Optional<Connector> pageantSelector() {
+    try {
+      if (!config.allows(AuthConfig.Option.PAGEANT)) {
+        logger.info("pageant option disabled");
+        return Optional.empty();
+      } else if (PageantConnector.isConnectorAvailable()) {
+        logger.info("pageant available");
+        return Optional.of(new PageantConnector());
+      } else {
+        logger.info("pageant not available");
+        return Optional.empty();
+      }
+    } catch (Throwable e) {
+      logger.info("ssh-agent could not be configured: {}", e.getMessage());
+      logger.debug("ssh-agent failure details.", e);
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Choose which socket factory to use.
+   *
+   * @return a working socket factory or {@code null} if none is available
+   */
+  private Optional<USocketFactory> determineUSocketFactory() {
+    return Stream.<Supplier<Optional<USocketFactory>>>of(this::ncSelector, this::jnaSelector)
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .findFirst();
+  }
+
+  private Optional<USocketFactory> jnaSelector() {
+    try {
+      return Optional.of(new JNAUSocketFactory());
+    } catch (Throwable e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1181)